test/luatest_helpers/server.lua

```diff
@@ -343,6 +344,31 @@ function Server:grep_log(what, bytes, opts)
     return found
 end
 
+function Server:assert_log_exactly_once(what, opts)
```

Comment:
The name "assert" usually means a one time thing in all the tests. But the implementation actually is waiting. I would suggest to find another name, which would reflect that more explicitly. Like `ensure` or `wait`.